text/0022-lazy-electron-download.md

```diff
@@ -0,0 +1,180 @@
+# RFC Template
+
+- Start Date: 2025-09-18
+- RFC PR: [electron/rfcs#0022](https://github.com/electron/rfcs/pull/0022)
+- Electron Issues:
+- Reference Implementation:
+- Status: **Proposed**
+
+# Lazy Electron Download (Remove postinstall Script)
+
+## Summary
+
+Remove the `postinstall` script from the `electron` npm package and instead download the Electron binary on-demand when the CLI is first spawned, improving security by allowing users to disable `postinstall` scripts while maintaining the current user experience.
+
+## Motivation
+
+The `postinstall` script in npm packages is a known attack vector in the npm ecosystem. Malicious packages can execute arbitrary code during installation, potentially compromising developer machines and CI/CD pipelines. Many security-conscious developers and organizations disable `postinstall` scripts entirely using `npm install --ignore-scripts` or `.npmrc` configuration. PNPM actually disables these scripts by default.
+
+Currently, the Electron npm package downloads the Electron binary during the `postinstall` phase. While this provides a good initial experience, it forces users who want to disable `postinstall` scripts to either:
+1. Accept the security risk of enabling postinstall scripts
+2. Manually handle Electron binary downloads through custom tooling
+3. Individually allowlist the Electron package (when this is supported)
+
+By moving the download to occur lazily when the `electron` CLI is first invoked, we can:
+- Allow security-conscious users to completely disable `postinstall` scripts
+- Maintain a transparent experience for most users (binary downloads when they first run `electron`)
+- Reduce the attack surface for supply chain attacks
+
+## Guide-level explanation
+
+For most Electron users, this change will be completely transparent. Instead of the Electron binary downloading when you run `npm install electron`, it will download the first time you run the `electron` command:
+
+```bash
+npm install electron
+# No download happens during installation
+
+npx electron .
+# Electron binary downloads here on first run
+# Subsequent runs use the cached binary
+```
+
+The download happens automatically and includes a progress indicator, just like the current `postinstall` experience.
+
+### For users who disable postinstall scripts
+
+If you currently disable `postinstall` scripts for security reasons (`npm install --ignore-scripts` or `ignore-scripts=true` in `.npmrc`), Electron will now work out of the box:
+
+```bash
+# In your .npmrc
+ignore-scripts=true
+
+npm install electron
+# No download, no errors
+
+npx electron .
+# Downloads binary on first run
+# Everything works as expected
+```
+
+### For edge cases requiring immediate download
+
+Some use cases require the Electron binary to be available immediately after installation (e.g., custom build scripts, Docker container builds). For these scenarios, we should introduce a new `install-electron` CLI command:
```

Comment:
I wonder if it'd be better to define a `bin` script in the `electron` module called `install-electron`?

This would avoid needing a separate package, and I think it would also simplify the `install-electron` logic, because now the `install-electron` script would be *inside* the `electron` package, so it wouldn't need to deal with package hoisting or Yarn workspaces or anything, because it would always know exactly where the `electron` package is installed.

With this approach, people could still run:

```bash
npm install electron
npx install-electron
```

Or run the bin command in their own postinstall:

```json
{
  "scripts": {
    "postinstall": "install-electron"
  }
}
```

Reply:
To clarify it would be a bin command in the same electron package.

npx runs any bin command, not just package names

Reply:
I wonder if we should reserve the `install-electron` package name so people don't run `npx install-electron` in the wrong dir and run some random untrusted code by accident.

Reply:
Good call, it's mine 😅 we can do what we want with it